### DISCUSSION TITLE
SoftI2CMaster with megaTinyCore

### DISCUSSION THREAD

**Discussion by TheFlav** [felias-fogg/SoftI2CMaster#74] [Ideas]
I started a discussion at https://github.com/SpenceKonde/megaTinyCore/discussions/664 that describes this.  I don't think I need to duplicate that here, but I'll mention the basics.

I wanted to use SoftI2CMaster on an attiny1627 chip (and some other similar parts).  It gave me some errors.  I came up with a workaround that I think isn't really a good fix.  In limited testing, it seems to work fine, but I had to edit the SoftI2CMaster.h file.

I think someone could implement the workaround as a real fix.  I just don't really know how to do that.  To that end, though, I will post some info here.

At https://www.microchip.com/en-us/product/ATTINY1627 there is a datasheet.  I think the relevant info applies to all the attiny 0, 1, and 2-series parts.  (https://ww1.microchip.com/downloads/aemDocuments/documents/MCU08/ProductDocuments/DataSheets/ATtiny1624-26-27-DataSheet-DS40002234B.pdf)  In section 8.1, there is a Peripheral Address Map.  Notice that VPORTA, VPORTB, and VPORTC are there at the top at address 0x00, 0x04, and 0x08.

Then, take a look at https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/ioheaders/iotn1627.h#L2181 and see how VPORTA_DIR, VPORTA_OUT, and VPORTA_IN are defined.  

Basically, my workaround was to modify SoftI2CMaster.h and add a #if around the _DDR, _OUT, and _IN defines like this.
```
#if !(defined(SDA_DDR) && defined(SDA_OUT) && defined(SDA_IN) && defined(SCL_DDR) && defined(SCL_OUT) && defined(SCL_IN))
#define SDA_DDR         (_SFR_IO_ADDR(SDA_PORT) - 1)
#define SCL_DDR         (_SFR_IO_ADDR(SCL_PORT) - 1)
#define SDA_OUT         _SFR_IO_ADDR(SDA_PORT)
#define SCL_OUT         _SFR_IO_ADDR(SCL_PORT)
#define SDA_IN      (_SFR_IO_ADDR(SDA_PORT) - 2)
#define SCL_IN      (_SFR_IO_ADDR(SCL_PORT) - 2)
#endif
```

And then add them to my .INO file like this.
```
#define SDA_DDR         0x00 //VPORTA_DIR = 0x00, VPORTB_DIR = 0x04, VPORTC_DIR = 0x08
#define SCL_DDR         0x00 //VPORTA_DIR = 0x00, VPORTB_DIR = 0x04, VPORTC_DIR = 0x08
#define SDA_OUT         0x01 //VPORTA_OUT = 0x01, VPORTB_OUT = 0x05, VPORTC_OUT = 0x09
#define SCL_OUT         0x01 //VPORTA_OUT = 0x01, VPORTB_OUT = 0x05, VPORTC_OUT = 0x09
#define SDA_IN          0x02 //VPORTA_IN  = 0x02, VPORTB_IN  = 0x06, VPORTC_IN  = 0x0A
#define SCL_IN          0x02 //VPORTA_IN  = 0x02, VPORTB_IN  = 0x06, VPORTC_IN  = 0x0A

#define SDA_PIN 5
#define SCL_PIN 4
#define I2C_FASTMODE 1

#include <SoftI2CMaster.h>
```

Which allows me to explicitly set them instead of using SDA_PORT and SCL_PORT.  I bet that there's a better way to do this for a library.  I just don't really know how to properly implement this for mass consumption.

Thanks!
Ed

**Comment by felias-fogg** [comment]
Hi Ed,

thanks for pointing out the problem. I must admit that I have not used one of the modern AVR MCUs. So, it may take a while until I come up with a solution. 

Best, 
Bernhard
